---
title: "DCI Bioinformatics tools: Summary statistics used for QC automatic report"
author: "Zhecheng Sheng"
date: "`r Sys.Date()`"
output: pdf_document
vignette: >
  %\VignetteIndexEntry{DCI Bioinformatics tools: Summary statistics used for QC automatic report}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE,
  echo = FALSE,
  results = "asis",
  fig.show = "hold",
  fig.width = 10,
  fig.height = 5,
  fig.align = 'center',
  tidy=TRUE, 
  tidy.opts=list(width.cutoff=60)
)
```

This documentation demonstrates the utility tools included in this package for generating automatic QC reports. Generally it takes data from STAR output and Picard output. If you want to report any issues using that package, please contact [zhecheng.sheng@duke.edu](zhecheng.sheng@duke.edu).

## Introduction


This is the package consists of a number of utility functions developed and maintained by the Duke Cancer Institute (DCI) Bioinformatics Group to facilitate reproducible bioinformatics analyses. This package includes 5 sample folders, storing the output for STAR run and Picard run, be aware that your directory structure may be different from this 5 sample files. 


## Workflow Using Example Data
```{r setup}
library(dcibioinformatics)
library(DESeq2)
```

 **Note**
 
 -  You can use `printab(table)` function in this package to present latex table in your report. 

### Tables output

* **Summary of STAR run**

```{r star summary, echo=TRUE}
rootdir <- dirname(system.file("extdata/samples","sample1/",
                package = "dcibioinformatics",mustWork = TRUE)) 

libnames <- list.files(rootdir)
filenames <- list.files(rootdir,pattern = "Log.final.out",recursive = T,full.names = T)
starsumm <- getSTARsum(libnames = libnames,files = filenames,verbose = F)
knitr::kable(starsumm[1:10,1:3],align = 'c',caption = 'STAR Alignment Summary')
```

* **Summary of STAR Counts**

```{r star counts,echo = TRUE}
files = list.files(rootdir,pattern = "ReadsPerGene.out.tab",recursive = T,full.names = T)
starcounts <- getSTARcounts(files = files,libnames = libnames,strand = "second")
knitr::kable(starcounts[1:10,],align = 'c',caption = 'STAR Counts Summary')
```

* **Summary of Picard Counts**
```{r picard counts, echo = TRUE}
files = list.files(rootdir,pattern = "metrics",recursive = T,full.names = T)
picounts <- getPIcounts(files,libnames)
knitr::kable(picounts,align = 'c',caption = 'Picard Counts Summary')
```

### Costructing RNAqc class

```{r Create RNAqc class, echo = TRUE, eval=TRUE, results='markup'}
coldata <- data.frame(condition = c('A','A','A','B','B'),genotype = c('W','P','W','P','W'),
row.names = paste0('sample',1:5), stringsAsFactors = TRUE)
### Create RNAqc instance
annodir <- dirname(system.file("extdata/","samples",package = "dcibioinformatics",mustWork = TRUE))
anno <- read.table(file.path(annodir, "annotation.txt"),header = TRUE,sep ='\t',stringsAsFactors = F)
rnaqc <- RNAqc(counts = starcounts, colData = coldata, picard = picounts, anno = anno)
show(rnaqc)
#### Change design
design(rnaqc) <- formula(~condition)

#### filter zero counts
rnaqc <- rnaqc[rowSums(DESeq2::counts(rnaqc))>0,]

#### Pre-normalize the 
rnaqc <- normMatrix(rnaqc)
```

**NOTE**
 
 - The annotation above is used mainly for demo purposes. 
 
 - In practice, please use the code below with actual gtf file.

#### Add corresponding annotation from gtf to the object.
```{r eval = FALSE, echo = TRUE}
library(rtracklayer)
gtf <- "path/to/gtf"
gtfObj <- rtracklayer::import(gtf)
rnaqc <- addGTF(rnaqc, gtfObj, id = "gene_id", type = "gene")
```

**Summary table for STAR mapping statistics**
```{r summary, eval =TRUE, echo = TRUE, results='markup'}
summary(rnaqc)
```


### Figures output

* **Counting Reads(STAR)**

```{r counting reads bar plot, echo=TRUE}
plot(rnaqc, type = "STAR")
```

* **Counting Bases(Picard)**

```{r counting bases bar plot, echo = TRUE}
#### Without group
plot(rnaqc, type = "PICARD",textsize = 5)
```

* **Unique Mapping rate**
```{r unique mapping rate, echo = TRUE}
plot(rnaqc, type = "MAPR",group = "condition")
```

### Basic RNASeq Analsis

* **PCA plot**
```{r pca plot, echo = T, message=F}
plot(rnaqc, type = "PCA", var = "condition", label = TRUE)
```
* **PCA plot (align PC axis)**
```{r pca plot align, echo = T, message=F}
plot(rnaqc, type = "PCA", var = "condition", label = TRUE, allocation = 'align')
```
* ** Only certain comparison **
```{r echo = T, message=F}
plot(rnaqc,type = "PCA", var = "condition", label = TRUE, pc = "p1")
```

* **Plot Size Factors**
```{r sizefactors ,echo = T, message=F}
plot(rnaqc, type = "SizeFactor", group = "condition")
```


* **Check expression level for selected genes (with box)**

```{r visualize normlized counts for selected gene boxplot, message=F,echo=T}
genelist = c("XWRnD","5hRk2","kfR6w","XGUbU")
plot(rnaqc, type = "Expression",group = "condition", target = genelist,gridvar = "genotype")
```
* **Check expression level for selected genes (without box)**

```{r visualize normlized counts for selected gene ,message=F,echo=T}
plot(rnaqc, type = "Expression",group = "condition", target = genelist,gridvar = "genotype", boxplot= FALSE, alpha = 0.5, shape = 2)
```
* **Check expression level without annotation** 
```{r eval = FALSE, echo = T}
plot(rnaqc,type = "Expression",group = "condition", target = c("gene1.1","gene2.1","gene3.1","gene4.1"),gridvar = "genotype")
```

### Run DESeq2 to get DEG results
```{r echo = T}
#### Use deseq2 simlulation to get significant visualization
dds = makeExampleDESeqDataSet(n = 2000,betaSD = 0.4, interceptMean = 5)
mcols(dds) = mcols(rnaqc)
mcols(dds)$gene_id = gsub('\\.\\d+','',mcols(dds)$gene_id)
#########
dds = as(dds, "RNAqc")
dds = DESeq2::DESeq(dds)
plot(dds, type = "Volcano")
```

### To Be Continued...

### Reference

-  [Extending the SummarizedExperiment class](http://bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/Extensions.html)


```{r echo = TRUE, results='markup'}
sessionInfo()
```