\nonstopmode{}
\documentclass[letterpaper]{book}
\usepackage[times,inconsolata,hyper]{Rd}
\usepackage{makeidx}
\usepackage[utf8]{inputenc} % @SET ENCODING@
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge Package `dcibioinformatics'}}
\par\bigskip{\large \today}
\end{center}
\begin{description}
\raggedright{}
\inputencoding{utf8}
\item[Title]\AsIs{Duke Cancer Institute (DCI) R Package}
\item[Version]\AsIs{0.1}
\item[Description]\AsIs{This package consists of a number of utility functions developed and maintained by the Duke Cancer Institute (DCI) Bioinformatics Group to facilitate reproducible bioinformatics analyses.}
\item[Depends]\AsIs{R (>= 3.6.1)}
\item[License]\AsIs{GPL-3 + file LICENSE}
\item[Encoding]\AsIs{UTF-8}
\item[LazyData]\AsIs{true}
\item[Imports]\AsIs{Rcpp (>= 1.0.1)}
\item[LinkingTo]\AsIs{Rcpp, RcppEigen}
\item[Suggests]\AsIs{knitr}
\item[VignetteBuilder]\AsIs{knitr}
\item[BuildVignettes]\AsIs{yes}
\item[NeedsCompilation]\AsIs{yes}
\item[RoxygenNote]\AsIs{6.1.1}
\item[Author]\AsIs{Kouros Owzar [aut, cre]}
\item[Maintainer]\AsIs{Kouros Owzar }\email{kouros.owzar@duke.edu}\AsIs{}
\end{description}
\Rdcontents{\R{} topics documented:}
\inputencoding{utf8}
\HeaderA{countCombine}{countCombine It combines count columns of different samples}{countCombine}
%
\begin{Description}\relax
countCombine
It combines count columns of different samples
\end{Description}
%
\begin{Usage}
\begin{verbatim}
countCombine(df1, df2)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{df1}] The strand-specific star counts for sample 1

\item[\code{df2}] The strand-specific star counts for sample 2
\end{ldescription}
\end{Arguments}
%
\begin{Value}
The joint df from df1 and df2
\end{Value}
\inputencoding{utf8}
\HeaderA{dcibioinformatics}{dcibinformatics.}{dcibioinformatics}
\aliasA{dcibioinformatics-package}{dcibioinformatics}{dcibioinformatics.Rdash.package}
%
\begin{Description}\relax
dcibinformatics.
\end{Description}
\inputencoding{utf8}
\HeaderA{getPIcounts}{getPIcounts Extract counting bases summary from PICARD output}{getPIcounts}
%
\begin{Description}\relax
getPIcounts
Extract counting bases summary from PICARD output
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getPIcounts(rootdir)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{rootdir}] Directory that stores all picard \code{.metrics} file
\end{ldescription}
\end{Arguments}
%
\begin{Value}
joint table for counting bases from PICARD output
\end{Value}
\inputencoding{utf8}
\HeaderA{getSTARcounts}{getSTARcounts Extract strand-specific STAR counts for each sample and then merge all counts into a single table}{getSTARcounts}
%
\begin{Description}\relax
getSTARcounts
Extract strand-specific STAR counts for each sample and then merge all counts into a single table
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getSTARcounts(rootdir, strand)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{rootdir}] The root directory under which the count file is stored

\item[\code{strand}] The strand-specific protocol used in RNA sequencing. Choose one from \code{c("first", "second", "unstranded")}
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A strand-specific STAR count summary table
\end{Value}
\inputencoding{utf8}
\HeaderA{getSTARsum}{getSTARsum Produce a summary table for STAR alignment process}{getSTARsum}
%
\begin{Description}\relax
getSTARsum
Produce a summary table for STAR alignment process
\end{Description}
%
\begin{Usage}
\begin{verbatim}
getSTARsum(rootdir)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{rootdir}] The root directory under which the count file is stored
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A joint table for STAR alignment staitistics
\end{Value}
\inputencoding{utf8}
\HeaderA{logCombine}{logCombine It combines log columns of different samples}{logCombine}
%
\begin{Description}\relax
logCombine
It combines log columns of different samples
\end{Description}
%
\begin{Usage}
\begin{verbatim}
logCombine(df1, df2)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{df1}] The star log file for sample 1

\item[\code{df2}] The star log file sample 2
\end{ldescription}
\end{Arguments}
%
\begin{Value}
The joint df from df1 and df2
\end{Value}
\inputencoding{utf8}
\HeaderA{myPidfile}{myPidfile get picard \code{.metrics} files from root directory}{myPidfile}
%
\begin{Description}\relax
myPidfile
get picard \code{.metrics} files from root directory
\end{Description}
%
\begin{Usage}
\begin{verbatim}
myPidfile(rootdir)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{rootdir}] Directory that stores all picard \code{.metrics} file
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A list contains all the \code{.metrics} files
\end{Value}
\inputencoding{utf8}
\HeaderA{myrcpp}{myrcpp The function illustrates how to use RcppEigen to fit the model Y = X beta + epsilon What is returned is the vector betahat=LSE(beta)}{myrcpp}
%
\begin{Description}\relax
myrcpp
The function illustrates how to use RcppEigen to fit the model
Y = X beta + epsilon
What is returned is the vector betahat=LSE(beta)
\end{Description}
%
\begin{Usage}
\begin{verbatim}
myrcpp(Y, X)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{Y}] the vector of response

\item[\code{X}] the matrix of covariates (does not include intercept)
\end{ldescription}
\end{Arguments}
%
\begin{Value}
\code{betahat}
\end{Value}
%
\begin{Author}\relax
Kouros Owzar
\end{Author}
\inputencoding{utf8}
\HeaderA{mystarfile}{mystarfile Generates the full path to a STAR count file. It assume that the count file is of the form rootdir/stardir/suffix}{mystarfile}
%
\begin{Description}\relax
mystarfile
Generates the full path to a STAR count file.
It assume that the count file is of the form
rootdir/stardir/suffix
\end{Description}
%
\begin{Usage}
\begin{verbatim}
mystarfile(rootdir, stardir, suffix = "ReadsPerGene.out.tab")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{rootdir}] The root directory under which the STAR output is stored

\item[\code{stardir}] The directory holding the files generated by STAR

\item[\code{suffix}] Suffix used by STAR to identify countfile
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Full path to the STAR count file
\end{Value}
\inputencoding{utf8}
\HeaderA{mystarLogfile}{mystarLogfile Generates the full path to a STAR log file. It assume that the log file is of the form rootdir/stardir/suffix}{mystarLogfile}
%
\begin{Description}\relax
mystarLogfile
Generates the full path to a STAR log file.
It assume that the log file is of the form
rootdir/stardir/suffix
\end{Description}
%
\begin{Usage}
\begin{verbatim}
mystarLogfile(rootdir, stardir, suffix = "Log.final.out")
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{rootdir}] The root directory under which the log file is stored

\item[\code{stardir}] The directory holding the files generated by STAR

\item[\code{suffix}] Suffix used by STAR to identify log file
\end{ldescription}
\end{Arguments}
%
\begin{Value}
Full path to the STAR log file
\end{Value}
\inputencoding{utf8}
\HeaderA{picardCombine}{picardCombine Row bind of picard summary for df1 and df2}{picardCombine}
%
\begin{Description}\relax
picardCombine
Row bind of picard summary for df1 and df2
\end{Description}
%
\begin{Usage}
\begin{verbatim}
picardCombine(df1, df2)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{df1}] picard summary for sample 1

\item[\code{df2}] picard summary for sample 2
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A joint summary dataframe from df1 and df2
\end{Value}
\inputencoding{utf8}
\HeaderA{plotAlignment}{plotAlignment Plot stacked barplot for bases counting proportion for STAR and PICARD}{plotAlignment}
%
\begin{Description}\relax
plotAlignment
Plot stacked barplot for bases counting proportion for STAR and PICARD
\end{Description}
%
\begin{Usage}
\begin{verbatim}
plotAlignment(counts, type = "HTSeq", textsize = 2, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{counts}] Summary counts table from getSTARcounts or getPIcounts

\item[\code{type}] Counting reads or counting bases, select one from \code{c("HTSeq","Picard")}

\item[\code{textsize}] Annotation size on the plot
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A ggplot type plot
\end{Value}
\inputencoding{utf8}
\HeaderA{printab}{printab Print summary table in LaTex format}{printab}
%
\begin{Description}\relax
printab
Print summary table in LaTex format
\end{Description}
%
\begin{Usage}
\begin{verbatim}
printab(tab, cap = NULL, top = 20, scale = 0.7, align = NULL, ...)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{tab}] Summary table from getSTARcounts, getSTARsum or getPIcounts

\item[\code{cap}] Character vector of length 1 or 2 containing the table's caption or title. If length is 2, the second item is the "short caption" used when LaTeX generates a "List of Tables". Set to NULL to suppress the caption. Default value is NULL

\item[\code{top}] number of rows to display, usually samples

\item[\code{scale}] parameters used to control the scale of the table

\item[\code{align}] Character vector of length equal to the number of columns of the resulting table, indicating the alignment of the corresponding columns. Also, "|" may be used to produce vertical lines between columns in LaTeX tables, but these are effectively ignored when considering the required length of the supplied vector. If a character vector of length one is supplied, it is split as strsplit(align, "")[[1]] before processing. Since the row names are printed in the first column, the length of align is one greater than ncol(x) if x is a data.frame. Use "l", "r", and "c" to denote left, right, and center alignment, respectively. Use "p3cm" etc. for a LaTeX column of the specified width. For HTML output the "p" alignment is interpreted as "l", ignoring the width request. Default depends on the class of x.
\end{ldescription}
\end{Arguments}
%
\begin{Value}
output table in LaTex format
\end{Value}
\inputencoding{utf8}
\HeaderA{sanitize\_text}{sanitize\_text Sanitize LaTex table output from xtable}{sanitize.Rul.text}
%
\begin{Description}\relax
sanitize\_text
Sanitize LaTex table output from xtable
\end{Description}
%
\begin{Usage}
\begin{verbatim}
sanitize_text(text)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{text}] LaTex table
\end{ldescription}
\end{Arguments}
%
\begin{Value}
sanitized table
\end{Value}
\inputencoding{utf8}
\HeaderA{sumHTSeq}{sumHTSeq Helper function that makes adjustments to output from getSTARcounts to be ready for plotting}{sumHTSeq}
%
\begin{Description}\relax
sumHTSeq
Helper function that makes adjustments to output from getSTARcounts to be ready for plotting
\end{Description}
%
\begin{Usage}
\begin{verbatim}
sumHTSeq(counts)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{counts}] Counts table from STAR alignment
\end{ldescription}
\end{Arguments}
%
\begin{Value}
A dataframe ready for plotting
\end{Value}
\printindex{}
\end{document}
